pacman/operations/routing_info_allocator_algorithms/zoned_routing_info_allocator.py

```diff
@@ -16,154 +16,308 @@
+                    return  # working result found
+            self.__n_bits_atoms -= 1
+
+        raise PacmanRouteInfoAllocationException(
+            "Unable to use ZonedRoutingInfoAllocator please select a "
+            "different allocator")
+
+    def __check_fixed(self):
+        self.__fixed_used = set()
+        bucket_size = 2 ** self.__n_bits_atoms_and_mac
+        for (key, n_keys) in self.__fixed_keys:
+            start = key // bucket_size
+            end = (key + n_keys) // bucket_size
+            for i in range(start, end):
+                self.__fixed_used.add(i)
+
+    def __allocate(self):
         """
-        :param int max_app_keys_bits: max bits for app keys
-        :param dict(ApplicationVertex,int) key_bits_map:
-            map of app vertex to max keys for that vertex
+        :param dict((ApplicationVertex, str),int) mask_bits_map:
+            map of app vertex,name to max keys for that vertex
         :return: tuple of routing infos and map from app vertex and key masks
-        :rtype: tuple(RoutingInfo, dict(ApplicationVertex,BaseKeyAndMask))
+        :rtype: tuple(RoutingInfo,
+            dict((ApplicationVertex, str), BaseKeyAndMask))
         """
+        by_app_and_partition_name = \
+            self.__machine_graph.multicast_partitions
         progress = ProgressBar(
-            self.__application_graph.n_vertices, "Allocating routing keys")
+            len(by_app_and_partition_name), "Allocating routing keys")
         routing_infos = RoutingInfo()
-        by_app_vertex = dict()
-        app_mask = self.__mask(max_app_keys_bits)
-
-        for source_index, app_vertex in progress.over(
-                enumerate(self.__application_graph.vertices)):
-            app_key = source_index << max_app_keys_bits
-            if app_vertex in key_bits_map:
-                key_bits = key_bits_map[app_vertex]
-                for machine_index, vertex in enumerate(
-                        app_vertex.machine_vertices):
-                    partitions = self.__machine_graph.\
-                        get_outgoing_edge_partitions_starting_at_vertex(vertex)
-                    partition = partitions.peek()
-                    if partition.traffic_type == EdgeTrafficType.MULTICAST:
-                        mask = self.__mask(key_bits)
-                        key = app_key | machine_index << key_bits
-                        key_and_mask = BaseKeyAndMask(base_key=key, mask=mask)
-                        routing_infos.add_partition_info(
-                            PartitionRoutingInfo([key_and_mask], partition))
-                by_app_vertex[app_vertex] = BaseKeyAndMask(
-                    base_key=app_key, mask=app_mask)
-
-        return routing_infos, by_app_vertex
+        app_part_index = 0
+        for app_id in progress.over(by_app_and_partition_name):
+            while app_id in self.__fixed_used:
+                app_part_index += 1
+            for partition_name, by_partition_name in \
```

Comment:
See previous comment re: naming

Reply:
Choose not to rename due to there being TWO levels.

Happy to rename if two names are suggested.

Reply:
I am happy with "partition_name", just "by_partition_name", which at this point is hard to trace... it is machine_vertices I believe; they are not "by_partition_name" at this point because you already have the partition_name here.

Reply:
renamed